tests/unit/client.py

```diff
@@ -0,0 +1,635 @@
+# Copyright 2013-2014 Massachusetts Open Cloud Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the
+# License.  You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an "AS
+# IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied.  See the License for the specific language
+# governing permissions and limitations under the License.
+
+"""Unit tests for client library"""
+import haas
+from haas import model, api, deferred, server, config
+from haas.model import db
+from haas.network_allocator import get_network_allocator
+import pytest
+import json
+import requests
+import os
+import tempfile
+import subprocess
+import time
+from subprocess import check_call, Popen
+from urlparse import urljoin
+import requests
+from requests.exceptions import ConnectionError
+from haas.client.base import ClientBase, FailedAPICallException
+from haas.client.client import Client, RequestsHTTPClient, KeystoneHTTPClient
+
+
+ep = "http://127.0.0.1:8000" or os.environ.get('HAAS_ENDPOINT')
+username = "hil_user" or os.environ.get('HAAS_USERNAME')
+password = "hil_pass1234" or os.environ.get('HAAS_PASSWORD')
+
+http_client = RequestsHTTPClient()
+http_client.auth = (username, password)
+C = Client(ep, http_client)  # Initializing client library
+MOCK_SWITCH_TYPE = 'http://schema.massopencloud.org/haas/v0/switches/mock'
+OBM_TYPE_MOCK = 'http://schema.massopencloud.org/haas/v0/obm/mock'
+OBM_TYPE_IPMI = 'http://schema.massopencloud.org/haas/v0/obm/ipmi'
+
+
+class Test_ClientBase:
+    """Tests client initialization and object_url creation. """
+
+    def test_init_error(self):
+        with pytest.raises(TypeError):
+            x = ClientBase()
+
+    def test_object_url(self):
+        x = ClientBase(ep, 'some_base64_string')
+        y = x.object_url('abc', '123', 'xy23z')
+        assert y == 'http://127.0.0.1:8000/abc/123/xy23z'
+
+# For testing the client library we need a running HIL server, with dummy
+# objects populated. Following classes accomplish that end.
+# It shall:
+#       1. Configures haas.cfg
+#       2. Instantiates a database
+#       3. Starts a server on an arbitary port
+#       4. Populates haas with dummy objects
+#       5. tears down the setup in a clean fashion.
+
+db_dir = None
+
+
+def make_config():
+    """ This function creates haas.cfg with desired options
+    and writes to a temporary directory.
+    It returns a tuple where (tmpdir, cwd) = ('location of haas.cfg', 'pwdd')
```

Comment:
Looks like your configuration has some problems.
Is the server running on port 5000 or port 5050

`DEBUG:requests.packages.urllib3.connectionpool:http://127.0.0.1:5050`
or
`Error: HTTPConnectionPool(host='127.0.0.1', port=5000): Max retries exceeded with url...`